File: code/java/src/ANN.java
import java.util.Random;

public class ANN {

    // Implements a 2-layer ANN

    public static final double initSigma = 0.02;
    public static final double relStopMargin = 0.01;
    public static int maxIterations = 1000000;
    public static double stepSize = 0.1;
    public Matrix hiddenWeights, outputWeights; // outputWeights are stored as a row matrix

    public static void main(String[] args) {
        ANN.test();
    }

    public ANN(int inputSize, int hiddenSize) {
        this.hiddenWeights = Matrix.createRandNorm(inputSize, hiddenSize, 0, ANN.initSigma);
        this.outputWeights = Matrix.createRandNorm(hiddenSize, 1, 0, ANN.initSigma);
    }

    // Helper methods

    public static double activation(double value) {
        return Math.tanh(value);
    }

    public static Vector activation(Vector vector) {
        // Applies the activation function to an entire vector in-place
        for(int i = 0; i < vector.height; i++) {
            vector.values[i] = ANN.activation(vector.values[i]);
        }
        return vector;
    }

    public static double dactivation(double value) {
        return 1 - Math.pow(Math.tanh(value), 2);
    }

    public static double frobNorm(Matrix a, Matrix b) {
        // Calculates the total Frobenius-norm of both matrices a and b
        return Math.sqrt(Math.pow(a.frobNorm(), 2) + Math.pow(b.frobNorm(), 2));
    }

    // ANN methods

    public double predict(Vector input) {

        // Predicts the output of this input vector
        // input is always normalized in this call

        Vector res = this.outputWeights.multiply(ANN.activation(this.hiddenWeights.multiply(input.normalize())));
        return ANN.activation(res.values[0]);

    }

    public void train(Example[] examples) {

        // Trains this ANN on the given array of examples
        // input vectors will always be normalized in this call

        long start = System.nanoTime();

        // Input normalization
        for(Example example : examples) {
            example.input.normalize();
        }

        // Pass over all examples until weights have converged enough
        int iteration = 0;
        while(true) {

            // Store old weights to check for convergence later
            Matrix prevHiddenWeights = this.hiddenWeights.deepcopy();
            Matrix prevOutputWeights = this.outputWeights.deepcopy();

            // Go through all examples
            for(Example example : examples) {

                // Calculate outputs
                Vector hiddenInput = this.hiddenWeights.multiply(example.input);
                Vector hiddenOutput = ANN.activation(hiddenInput.deepcopy());
                double finalInput = this.outputWeights.multiply(hiddenOutput).values[0];
                double predictedOutput = ANN.activation(finalInput);

                // Used in calculations
                double predictionError = example.output - predictedOutput;
                double outputDerivative = ANN.dactivation(finalInput);

                // Adjust output weights and calculate requested hidden change
                Matrix requestedHiddenChange = this.outputWeights.deepcopy().multiply(predictionError*outputDerivative); // Row matrix, like output weights
                for(int i = 0; i < this.outputWeights.width; i++)
                    this.outputWeights.values[i][0] += ANN.stepSize*predictionError*hiddenOutput.values[i];

                // Backpropagate requested hidden change to adjust hidden weights
                for(int i = 0; i < this.hiddenWeights.width; i++) {
                    for(int j = 0; j < this.hiddenWeights.height; j++) {
                        this.hiddenWeights.values[i][j] += ANN.stepSize*requestedHiddenChange.values[j][0]*ANN.dactivation(hiddenInput.values[j])*example.input.values[i];
                    }
                }

                // Check stop criteria
                iteration++;
                if (iteration >= ANN.maxIterations)
                    break;

            }

            // Check stop criteria
            if (iteration >= ANN.maxIterations)
                break;
            // Convergence check
            double diff = ANN.frobNorm(prevHiddenWeights.multiply(-1).add(this.hiddenWeights), prevOutputWeights.multiply(-1).add(this.outputWeights));
            double base = ANN.frobNorm(this.hiddenWeights, this.outputWeights);
            //if (base >= CustomMath.epsilon && diff/base < ANN.relStopMargin)
            //    break;

        }

        System.out.println((System.nanoTime() - start)/1000000000.0);
        System.out.println("Stopped training after " + iteration + " iterations.");

    }

    // Testing methods

    public static void test() {

        // Test the ANN by having it model another ANN with identical topology but unknown weights

        int inputSize = 5;
        int hiddenSize = 3;
        ANN real = new ANN(inputSize, hiddenSize);
        ANN model = new ANN(inputSize, hiddenSize);

        // Generate training data
        int amount = 10000;
        Example[] trainingSet = ANN.generateExamples(amount, inputSize, real);
        Example[] validationSet = ANN.generateExamples(amount, inputSize, real);

        // Print initial analysis, then train, then print new analysis
        ANN.printRMSEs(model, trainingSet, validationSet);
        model.train(trainingSet);
        ANN.printRMSEs(model, trainingSet, validationSet);

    }

    public static Example[] generateExamples(int amount, int inputSize, ANN real) {

        // Generates examples based on an ANN
        Example[] examples = new Example[amount];
        Random rand = new Random();
        for(int i = 0; i < amount; i++) {
            double[] inputValues = new double[inputSize];
            for(int j = 0; j < inputSize; j++)
                inputValues[j] = rand.nextGaussian();
            Vector input = new Vector(inputValues);
            examples[i] = new Example(input, real.predict(input));
        }
        return examples;

    }

    public static void printRMSEs(ANN model, Example[] trainingSet, Example[] validationSet) {

        // Calculates and prints the RMSE's of this model on the training and validation set
        System.out.println("RMSE on training data: " + ANN.RMSE(model, trainingSet));
        System.out.println("RMSE on validation data: " + ANN.RMSE(model, validationSet));

    }

    public static double RMSE(ANN ann, Example[] examples) {

        // Calculates the RMSE of this ann on this set of examples
        double total = 0;
        for(Example example : examples) {
            total += Math.pow(example.output - ann.predict(example.input), 2);
        }
        return Math.sqrt(total/examples.length);

    }

}

File: code/python/ann.py
from numpy import *
from math import sqrt
from copy import deepcopy
from time import time

class ANN:
	
	"""ANN with one hidden layer, one output and full connections in between consecutive layers.
	Initial weights are chosen from a normal distribution.
	Activation function is tanh."""
	
	INIT_SIGMA = 0.02
	REL_STOP_MARGIN = 0.01
	MAX_ITERATIONS = 1000000
	ACTIVATION = tanh
	D_ACTIVATION = lambda x: 1 - tanh(x)**2 # Derivative of tanh
	VEC_ACTIVATION = vectorize(ACTIVATION)
	VEC_D_ACTIVATION = vectorize(D_ACTIVATION)
	STEP_SIZE = 0.1
	
	def __init__(self, input_size, hidden_size):
		
		#self.input_size = input_size
		#self.hidden_size = hidden_size
		self.hidden_weights = random.normal(0, ANN.INIT_SIGMA, (hidden_size, input_size))
		self.output_weights = random.normal(0, ANN.INIT_SIGMA, hidden_size)
	
	def get_weights(self):
		return self.hidden_weights, self.output_weights
	
	def predict(self, input_vector):
		
		# Predicts the output for this input vector
		# input_vector will be normalized
		
		input_vector = input_vector/linalg.norm(input_vector)
		return ANN.ACTIVATION(dot(self.output_weights, ANN.VEC_ACTIVATION(dot(self.hidden_weights, input_vector))))
	
	@staticmethod
	def frob_norm(a, b):
		
		# Calculates the total Frobenius norm of both matrices A and B
		return sqrt(linalg.norm(a)**2 + linalg.norm(b)**2)
	
	def train(self, examples):
		
		#print("Training")
		start = time()
		
		# examples is a list of (input, output)-tuples
		# input will be normalized
		# We stop when the weights have converged within some relative margin
		
		for example in examples:
			example[0] = example[0]/linalg.norm(example[0])
		
		iteration = 0
		while True:
			
			
			# Store old weights to check for convergence later
			prev_hidden_weights = deepcopy(self.hidden_weights)
			prev_output_weights = deepcopy(self.output_weights)
			
			for k in range(len(examples)):
				
				input_vector, output = examples[k]
				
				# Calculate outputs
				hidden_input = dot(self.hidden_weights, input_vector)
				hidden_output = ANN.VEC_ACTIVATION(hidden_input)
				final_input = dot(self.output_weights, hidden_output)
				predicted_output = ANN.ACTIVATION(final_input)
				
				#print("Output:", output)
				#print("Predicted output:", predicted_output)
				
				# Used in calculations
				prediction_error = output - predicted_output
				output_derivative = ANN.D_ACTIVATION(final_input)
				
				# Adjust output weights and calculate requested hidden change
				requested_hidden_change = prediction_error*output_derivative*self.output_weights
				self.output_weights = self.output_weights + ANN.STEP_SIZE*prediction_error*hidden_output
				
				#print("After adjusting output weights:", ANN.ACTIVATION(dot(self.output_weights, hidden_output)))
				
				# Backpropagate requested hidden change to adjust hidden weights
				self.hidden_weights = self.hidden_weights + ANN.STEP_SIZE*outer(requested_hidden_change*(ANN.VEC_D_ACTIVATION(hidden_input)), input_vector)
				
				#print("After adjusting hidden weights:", ANN.ACTIVATION(dot(self.output_weights, ANN.VEC_ACTIVATION(dot(self.hidden_weights, input_vector)))))
				
				# Check stop criteria
				iteration += 1
				if iteration >= ANN.MAX_ITERATIONS:
					break
			
			# Check stop criteria
			if iteration >= ANN.MAX_ITERATIONS:
				break
			diff = ANN.frob_norm(self.hidden_weights - prev_hidden_weights, self.output_weights - prev_output_weights)
			base = ANN.frob_norm(self.hidden_weights, self.output_weights)
			#if base > 0 and diff/base < ANN.REL_STOP_MARGIN:
			#	break
		
		print(time() - start)
		print("Stopped training after %s iterations."%iteration)

# TESTING

def print_difference(ann1, ann2):
	
	# Prints the differences in weights in between two ANN's with identical topology
	
	hidden_weights1, output_weights1 = ann1.get_weights()
	hidden_weights2, output_weights2 = ann2.get_weights()
	hidden_diff = hidden_weights1 - hidden_weights2
	output_diff = output_weights1 - output_weights2
	
	print(hidden_diff)
	print(output_diff)
	print("Frobenius norms:")
	print("Hidden weights difference:", linalg.norm(hidden_diff))
	print("Output weights difference:", linalg.norm(output_diff))
	print("Both:", ANN.frob_norm(hidden_diff, output_diff))

def RMSE(ann, examples):
	
	total = 0
	for input_vector, output in examples:
		total += (output - ann.predict(input_vector))**2
	return sqrt(total/len(examples))

def generate_examples(amount, input_size, evaluate):
	# evaluate is a function mapping an input vector onto a numerical value
	examples = []
	inputs = random.normal(0, 100, (amount, input_size))
	for i in range(amount):
		input_vector = inputs[i]
		examples.append([input_vector, evaluate(input_vector)])
	return examples

def test():
	
	# Test the ANN by having it model another ANN with identical topology but unknown weights
	
	input_size = 5
	hidden_size = 3
	real = ANN(input_size, hidden_size)
	model = ANN(input_size, hidden_size)
	
	# Generate training data
	training_data = generate_examples(10000, input_size, real.predict)
	validation_data = generate_examples(10000, input_size, real.predict)
	
	# Print initial difference, train, then print new difference
	print("Initial difference:")
	print_difference(real, model)
	print("Initial RMSE (on training data):", RMSE(model, training_data))
	print("Initial RMSE (on validation data):", RMSE(model, validation_data))
	model.train(training_data)
	print("After training:")
	print_difference(real, model)
	print("After training RMSE (on training data):", RMSE(model, training_data))
	print("After training RMSE (on validation data):", RMSE(model, validation_data))

if __name__ == "__main__":
	test()


